src/Corax/Indexing/IndexWriter.cs

```diff
@@ -728,90 +773,88 @@ public void TryDeleteEntryByField(string field, string term)
                     return;
             }
 
-            RecordDeletion(idInTree, out _, out var setsAreDisjoint);
+            RecordAndPrepareDocumentsIdsForDeletion(idInTree, out var setsAreDisjoint, out _);
+            ProcessCurrentDeletes();
+            
             if (setsAreDisjoint == false)
             {
                 FlushBatch();
                 TryDeleteEntryByField(field, term);
             }
-            
         }
-        
+
         /// <summary>
         /// Record term for deletion from Index.
         /// </summary>
         /// <param name="idInTree">With frequencies and container type.</param>
-        /// <param name="setsAreDisjoint">Intersection between PostingList and _deletedEntries. We may use it as indicate for flushing batch.</param>
+        /// <param name="setsAreDisjoint">Intersection between PostingList and _deletedEntries. We may use it as indicator for flushing batch.</param>
         [SkipLocalsInit]
-        private bool RecordDeletion(long idInTree, out long singleEntryId, out bool setsAreDisjoint)
+        private void RecordAndPrepareDocumentsIdsForDeletion(long postingListId, out bool setsAreDisjoint, out bool isSingleDocument)
         {
+            Debug.Assert(_entriesToDelete.Count == 0);
+            
             var countOfAlreadyDeletedEntries = _deletedEntries.Count;
-            var totalRead = 0L;
             setsAreDisjoint = true;
-            var containerId = EntryIdEncodings.GetContainerId(idInTree);
-            if ((idInTree & (long)TermIdMask.PostingList) != 0)
+            var containerId = EntryIdEncodings.GetContainerId(postingListId);
+
+            if ((postingListId & (long)TermIdMask.EnsureIsSingleMask) == (long)TermIdMask.Single)
+            {
+                var singleEntryId = containerId;
+                Debug.Assert(singleEntryId > 0);
+                var isNewDocument = _deletedEntries.Add(containerId);
+                if (isNewDocument)
+                    _entriesToDelete.Add(singleEntryId);
+                setsAreDisjoint &= isNewDocument;
+                _numberOfModifications -= _deletedEntries.Count - countOfAlreadyDeletedEntries;
```

Comment:
Just to make sure - `_deletedEntries` is the hash set so it won't have duplicates. Is it okay that `_numberOfModifications` doesn't account for duplicates as well?